### ISSUE TITLE
NaiveDateTime::timestamp_nanos() panics.

### ISSUE THREAD

**Issue by mbr** [chronotope/chrono#310]
Minimal test case: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f83cb5d9ff1ac584805493612992327c

```rust
use chrono::NaiveDateTime;

fn main() {
    let problematic = "2262-04-11T23:47:16.855000054";
    let parsed: NaiveDateTime = problematic.parse().unwrap();
    println!("Problematic: {}, Parsed: {:?}", problematic, parsed);
    parsed.timestamp_nanos();
}
```

This panics:

> thread 'main' panicked at 'attempt to add with overflow', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.6/src/naive/datetime.rs:327:9

The resulting timestamp in nanoseconds is close to 63 bits, but not quite. If this is intended, the documentation should mention the possible overflow here?

**Comment by quodlibetor** [comment]
Ah yeah @mbr this is known, but poorly documented. I got the exact bounds and improved the documentation.

Without `i128` there isn't a lot that we can do about this, I don't think, and even with that I'm not sure what a nice API would look like.

Out of curiosity, is this a major issue for you? Can you talk about what use cases would make you want nanosecond precision ~200 years into the future?

Sorry about that annoyance!

**Comment by mbr** [comment]
> Ah yeah @mbr this is known, but poorly documented. I got the exact bounds and improved the documentation.

Awesome, I can't find the branch the fix is on though.

> Out of curiosity, is this a major issue for you? Can you talk about what use cases would make you want nanosecond precision ~200 years into the future?

I am writing a conversion function to fractional julian day dates; this is one of the way SQLite can store timestamp in the database (as double precision floats). Resolution, of course, is not nanosecond everywhere, it actually depends on when the date is. Around 1900-2100, it should be at roughly 80 microseconds, if I did my math right.

**Comment by quodlibetor** [comment]
The fix is in pr #311 if you feel like giving it a quick review.

Also I forgot to mention that since it is an i64 to go pre-1970 it does make sense that it panics 63 bits into the future.
